addons/nuxeo-core-binarymanager-cloud/nuxeo-core-binarymanager-s3/src/test/java/org/nuxeo/ecm/blob/s3/TestS3BlobStoreTracing.java

```diff
@@ -0,0 +1,522 @@
+
+    // from test-blob-provider-s3-tracing.xml, for cleanup
+    protected static final List<String> BLOB_PROVIDER_IDS = Arrays.asList( //
+            "s3", //
+            "s3-other", //
+            "s3-sha256-async", //
+            "s3-nocache", //
+            "s3-managed", //
+            "s3-record");
+
+    @Inject
+    protected LogCaptureFeature.Result logCaptureResult;
+
+    @Inject
+    protected BlobManager blobManager;
+
+    @Inject
+    protected TransactionalFeature txFeature;
+
+    protected Path tmpFile;
+
+    @BeforeClass
+    public static void beforeClass() {
+        S3TestHelper.getProperties().forEach(S3TestHelper::setProperty);
+    }
+
+    @Before
+    public void setUp() throws IOException {
+        clearBlobStores();
+        tmpFile = Files.createTempFile("tmp_", ".tmp");
+        logCaptureResult.clear();
+    }
+
+    @After
+    public void tearDown() throws IOException {
+        clearBlobStores();
+        Files.deleteIfExists(tmpFile);
+    }
+
+    protected void clearBlobStores() {
+        BLOB_PROVIDER_IDS.forEach(id -> getBlobStore(id).clear());
+    }
+
+    protected BlobStore getBlobStore(String id) {
+        return ((BlobStoreBlobProvider) getBlobProvider(id)).store;
+    }
+
+    protected BlobProvider getBlobProvider(String id) {
+        return blobManager.getBlobProvider(id);
+    }
+
+    protected void clearCache(BlobProvider blobProvider) throws IOException {
+        clearCache(((BlobStoreBlobProvider) blobProvider).store);
+    }
+
+    protected void clearCache(BlobStore blobStore) throws IOException {
+        if (blobStore instanceof TransactionalBlobStore) {
+            TransactionalBlobStore bs = (TransactionalBlobStore) blobStore;
+            clearCache(bs.store);
+            clearCache(bs.transientStore);
+        } else if (blobStore instanceof CachingBlobStore) {
+            CachingBlobStore bs = (CachingBlobStore) blobStore;
+            FileUtils.cleanDirectory(bs.cacheDir.toFile());
+        }
+    }
+
+    protected void clearTrace() {
+        logCaptureResult.clear();
+    }
+
+    protected void logTrace(String message) {
+        LogEvent event = Log4jLogEvent.newBuilder() //
+                                      .setMessage(new SimpleMessage(message))
+                                      .build();
+        logCaptureResult.getCaughtEvents().add(event);
+    }
+
+    protected void checkTrace(String filename) throws IOException {
+        List<String> expectedTrace = LogTracingHelper.readTrace("traces/" + filename);
+        List<String> actualTrace = logCaptureResult.getCaughtEventMessages();
+        Map<String, String> context = new HashMap<>();
+        try {
+            LogTracingHelper.assertEqualsLists(expectedTrace, actualTrace, context);
+        } catch (AssertionError e) {
+            System.err.println(filename);
+            System.err.println(String.join("\n", actualTrace));
```

Comment:
To replace by a logger?

Reply:
It's only for when the assertion fails, it's easier for me to have it on stderr.